server/src/main/java/org/elasticsearch/index/IndexSortConfig.java

```diff
@@ -95,30 +98,91 @@ public final class IndexSortConfig {
      */
     public static final Setting<List<String>> INDEX_SORT_MISSING_SETTING = Setting.listSetting(
         "index.sort.missing",
-        Collections.emptyList(),
+        IndexSortConfigDefaults::getDefaultSortMissing,
         IndexSortConfig::validateMissingValue,
         Setting.Property.IndexScope,
         Setting.Property.Final,
         Setting.Property.ServerlessPublic
     );
 
-    public static final FieldSortSpec[] TIME_SERIES_SORT, TIMESTAMP_SORT, HOSTNAME_TIMESTAMP_SORT, HOSTNAME_TIMESTAMP_BWC_SORT;
-    static {
-        FieldSortSpec timeStampSpec = new FieldSortSpec(DataStreamTimestampFieldMapper.DEFAULT_PATH);
-        timeStampSpec.order = SortOrder.DESC;
-        TIME_SERIES_SORT = new FieldSortSpec[] { new FieldSortSpec(TimeSeriesIdFieldMapper.NAME), timeStampSpec };
-        TIMESTAMP_SORT = new FieldSortSpec[] { timeStampSpec };
-
-        FieldSortSpec hostnameSpec = new FieldSortSpec(IndexMode.HOST_NAME);
-        hostnameSpec.order = SortOrder.ASC;
-        hostnameSpec.missingValue = "_last";
-        hostnameSpec.mode = MultiValueMode.MIN;
-        HOSTNAME_TIMESTAMP_SORT = new FieldSortSpec[] { hostnameSpec, timeStampSpec };
-
-        // Older indexes use ascending ordering for host name and timestamp.
-        HOSTNAME_TIMESTAMP_BWC_SORT = new FieldSortSpec[] {
-            new FieldSortSpec(IndexMode.HOST_NAME),
-            new FieldSortSpec(DataStreamTimestampFieldMapper.DEFAULT_PATH) };
+    public static class IndexSortConfigDefaults {
+        public static final FieldSortSpec[] TIME_SERIES_SORT, TIMESTAMP_SORT, HOSTNAME_TIMESTAMP_SORT, HOSTNAME_TIMESTAMP_BWC_SORT;
+
+        static {
+            FieldSortSpec timeStampSpec = new FieldSortSpec(DataStreamTimestampFieldMapper.DEFAULT_PATH);
+            timeStampSpec.order = SortOrder.DESC;
+            TIME_SERIES_SORT = new FieldSortSpec[] { new FieldSortSpec(TimeSeriesIdFieldMapper.NAME), timeStampSpec };
+            TIMESTAMP_SORT = new FieldSortSpec[] { timeStampSpec };
+
+            FieldSortSpec hostnameSpec = new FieldSortSpec(IndexMode.HOST_NAME);
+            hostnameSpec.order = SortOrder.ASC;
+            hostnameSpec.missingValue = "_last";
+            hostnameSpec.mode = MultiValueMode.MIN;
+            HOSTNAME_TIMESTAMP_SORT = new FieldSortSpec[] { hostnameSpec, timeStampSpec };
+
+            // Older indexes use ascending ordering for host name and timestamp.
+            HOSTNAME_TIMESTAMP_BWC_SORT = new FieldSortSpec[] {
+                new FieldSortSpec(IndexMode.HOST_NAME),
+                new FieldSortSpec(DataStreamTimestampFieldMapper.DEFAULT_PATH) };
+        }
+
+        public static FieldSortSpec[] getDefaultSortSpecs(Settings settings) {
+            if (settings.isEmpty()) {
+                return new FieldSortSpec[0];
+            }
+
+            String indexMode = settings.get(IndexSettings.MODE.getKey());
```

Comment:
Can't use `IndexSettings.MODE.get(settings)` here because the validation logic for `IndexSettings.MODE` uses the default value of `index.sort.*`, which causes infinite recursion (since we're already in the default value provider for those settings).
So we need to get the mode while bypassing the validation.

Reply:
Maybe add this also as a comment?